qiskit/transpiler/passes/basis/decompose.py

```diff
@@ -42,18 +55,46 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             output dag where ``gate`` was expanded.
         """
         # Walk through the DAG and expand each non-basis node
-        for node in dag.op_nodes(self.gate):
-            # opaque or built-in gates are not decomposable
-            if not node.op.definition:
-                continue
-            # TODO: allow choosing among multiple decomposition rules
-            rule = node.op.definition.data
-
-            if len(rule) == 1 and len(node.qargs) == len(rule[0][1]) == 1:
-                if node.op.definition.global_phase:
-                    dag.global_phase += node.op.definition.global_phase
-                dag.substitute_node(node, rule[0][0], inplace=True)
-            else:
-                decomposition = circuit_to_dag(node.op.definition)
-                dag.substitute_node_with_dag(node, decomposition)
+        for node in dag.op_nodes():
+            if self.should_decompose(node):
+                if not node.op.definition:
+                    continue
+                # TODO: allow choosing among multiple decomposition rules
+                rule = node.op.definition.data
+                if len(rule) == 1 and len(node.qargs) == len(rule[0][1]) == 1:
+                    if node.op.definition.global_phase:
+                        dag.global_phase += node.op.definition.global_phase
+                    dag.substitute_node(node, rule[0][0], inplace=True)
+                else:
+                    decomposition = circuit_to_dag(node.op.definition)
+                    dag.substitute_node_with_dag(node, decomposition)
+
         return dag
+
+    def should_decompose(self, node) -> bool:
```

Comment:
Do we want to keep this method public or private? If we keep it public, could you add some tests? Otherwise I'm good making it private 🙂